internal/openfga/stores.go

```diff
@@ -282,11 +282,92 @@ func (s *OpenFGAStore) ListPermissions(ctx context.Context, ID string, continuat
 	return permissions, tMap, fmt.Errorf(eMsg)
 }
 
-func (s *OpenFGAStore) listPermissionsFunc(ctx context.Context, ID, ofgaType, cToken string) func() any {
+// ListPermissionsWithFilters returns all the permissions associated to a specific entity
+func (s *OpenFGAStore) ListPermissionsWithFilters(ctx context.Context, ID string, opts ...ListPermissionsFiltersInterface) ([]Permission, map[string]string, error) {
+	ctx, span := s.tracer.Start(ctx, "openfga.OpenFGAStore.ListPermissionsWithFilters")
+	defer span.End()
+
+	// keep it a buffered channel, if set to unbuffered we would need a goroutine
+	// to consume from it before pushing to it
+	// https://go.dev/ref/spec#Send_statements
+	// A send on an unbuffered channel can proceed if a receiver is ready.
+	// A send on a buffered channel can proceed if there is room in the buffer
+	results := make(chan *pool.Result[any], len(s.permissionTypes()))
+
+	ff := new(listPermissionsOpts)
+
+	if len(opts) != 0 {
+		ff = s.parseFilters(opts...)
+	}
+
+	types := s.permissionTypes()
+	tokenMap := make(map[string]string)
+
+	if tm := ff.TokenMap; tm != nil {
+		tokenMap = tm
+	}
+
+	if tf := ff.TypesFilter; len(tf) > 0 {
+		types = tf
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(types))
+
+	for _, t := range types {
+		token, ok := tokenMap[t]
+
+		if !ok {
+			token = ""
+		}
+
+		s.wpool.Submit(
+			s.listPermissionsFunc(ctx, ID, ff.RelationFilter, t, token),
+			results,
+			&wg,
+		)
+	}
+
+	// wait for tasks to finish
+	wg.Wait()
+
+	// close result channel
+	close(results)
+
+	permissions := make([]Permission, 0)
+	tMap := make(map[string]string)
+	errors := make([]error, 0)
+
+	for r := range results {
+		v := r.Value.(listPermissionsResult)
+		permissions = append(permissions, v.permissions...)
+		tMap[v.ofgaType] = v.token
+
+		if v.err != nil {
+			errors = append(errors, v.err)
+		}
+	}
+
+	if len(errors) == 0 {
+		return permissions, tMap, nil
+	}
+
+	eMsg := ""
+
+	for n, e := range errors {
+		s.logger.Errorf(e.Error())
+		eMsg = fmt.Sprintf("%s%v - %s\n", eMsg, n, e.Error())
+	}
```

Comment:
this looks like something we could stick in the `pool` package for reuse

Reply:
ditto https://github.com/canonical/identity-platform-admin-ui/issues/384